### PR TITLE
firebase-jobdispatcher to built-in jobscheduler

### LINKED ISSUE #274: Notifications without GCM
It turns out there is a notifications service, which doesn't work if there are no Google Play Services.

- minSdkVersion 21 - so there is the standard [JobScheduler available](https://developer.android.com/reference/android/app/job/JobScheduler.html), why not use it?
- https://github.com/evernote/android-job is also a nice and flexible solution, more universal

Just for fun, I have already ported it to android-job :)
What do you think?
This is also important for #136

### PR DESCRIPTION
**Pull request by thermatk**
A better and easier approach than evernote, since minsdk is set to 21 anyway.
firebase-jobdispatcher in this case is just using the built-in one, so it's cutting out the middleman. If you would have minsdk<21, firebase or evernote would be using other routes available, GCM for firebase or GCM/AlarmManager for evernote.

Please test that there is no bug associated with GCM calling a class which is not for GCM anymore as mentioned here: https://github.com/evernote/android-job/blob/master/FAQ.md#how-can-i-remove-the-gcm-dependency-from-my-app . I don't know if it applies to firebase-jobdispatcher and/or api 21+, where GCM shouldn't have been used, as well.
Only you can do it, having the keys for both gcm and apk signing, so you can overwrite one version over the previous :+1: 
How to test:

1. Start app(release config with gcm) with current approach, an older build for the PlayStore should work
2. Install over it one built with this branch
3. Wait for the notification period to pass(30 minutes, or maybe even decrease it earlier on?)
4. See if there was a crash

Fixes #274

### CODE REVIEW

**Comment by k0shk0sh** [comment]
Thanks for the PR, FastHub currently does not use GCM/FCM at all & in fact, running the job is just to grab the latest notifications from GitHub which you can build & test it very easily, All you have to do is to follow the contribution guide to register for GitHub client & secret ID to be able to login & verify the results yourself, you could definitely decrease the notification interval from the settings screen to lets say one minute & then you could test out if the notification is received either via creating a dummy account in GitHub if you don't have a secondary one, or let me know and i'll buzz you on this PR to be able to receive notification. 
Let me know your thoughts.

**Comment by thermatk** [comment]
@k0shk0sh the possible(probably not existing, but still concerns me) is not with that. It's about when you used not the built-in JobScheduler of Android, but the GCM-based one. It looks like the removal/change of the associated service afterwards results in gcm still trying to start it anyway, crashing the app every time.

As for the service itself, it works fine here with my "debug" flavor and my own API keys.

**Comment by k0shk0sh** [comment]
Owh, you mean that the PlayService is actually cached the job & tries to start it every time? well that's weird. 
Have you got it run & it actually crashed the app?

**Comment by k0shk0sh** [comment]
Could you attach the logcat in here so I could have a look?

**Comment by thermatk** [comment]
No, that's the problem. For it to crash I should be able to install the new version over one where play services were active before. That would require a full release build of the original code with my own gcm/firebase keys + my own signing keys.

Can't test for it

**Comment by k0shk0sh** [comment]
Owwh, Alright, I got you, let me try that now & confirm with you.

**Comment by thermatk** [comment]
Otherwise, it's a pretty simple replacement of the abstraction layer in the middle, with a side effect being notifications will be working on devices without play services installed

**Comment by thermatk** [comment]
One day and notifications still going through :+1: 
If you can build this branch in release flavor, sign with your keys and share the apk, I can test it for the possible failure as well

**Comment by k0shk0sh** [comment]
@thermatk Oops, I literally forgot about it as i'm focusing on closing issues before today release.
Alright, I'll build it and share the apk with you.

Thanks man & I hope you forgive my lateness.

**Comment by k0shk0sh** [comment]
[app-release.apk.zip](https://github.com/k0shk0sh/FastHub/files/908181/app-release.apk.zip)

**Comment by thermatk** [comment]
@k0shk0sh not a problem!
So, over-night testing shows that firebase seems to have it fixed on their side. This is what shows hidden deep in logcat during installation:
`04-09 22:14:04.712 W/ActivityManager(2158): Scheduling restart of crashed service com.fastaccess.github/com.firebase.jobdispatcher.GooglePlayReceiver in 1000ms`

No FCs, no crashes and the firebase service never returned afterwards. The notification service, on the other hand, worked as expected, being run many times through the night :+1:

**Comment by k0shk0sh** [comment]
That's cool, I'll test it out more later today & onwards merge it. 
Thanks man. appreciate your effort really!

**Comment by thermatk** [comment]
test went well? :question:

**Comment by k0shk0sh** [comment]
Haven't confirm on yet, but doing it now.

**Comment by k0shk0sh** [comment]
Strangely, I can't receive any notifications. despite the fact that I set the interval to 1 minute.

**Comment by thermatk** [comment]
It fires fine here, but 15(not precisely, it's not the battery-killer AlarmManager 12-19) minutes minimum
you should've seen it in logcat:
`04-11 06:54:17.460 2290-2290/com.fastaccess.github W/JobInfo: Specified interval for 1 is +10m0s0ms. Clamped to +15m0s0ms`

**Comment by thermatk** [comment]
According to SO, this is a Nougat-specific optimization on the platform side:
http://stackoverflow.com/questions/38344220/job-scheduler-not-running-on-android-n

**Comment by k0shk0sh** [comment]
@thermatk job dispatcher will use job scheduler as well if > = 21 , but yet when I set the job interval to 1 minute I could receive the notification, however with this I couldn't really get it. Let me try again and get back to you.

**Comment by thermatk** [comment]
@k0shk0sh might be they are actually using gcm for it, I can build a workaround for <15minutes, to make it just as battery-killer as the previous one :+1: 
a job to run once in a minute and schedule another job to be run once in a minute?

**Comment by k0shk0sh** [comment]
Well, It took the job 11 minutes to fire up for a one minute scheduler. If we have the option for a one minute, then everybody expect that in one minute they should receive notifications. 

@thermatk yup, that would be great, anything that less than 15 minutes should have that workaround, also, could you handle this scenario? https://developer.android.com/reference/android/app/job/JobService.html#jobFinished(android.app.job.JobParameters, boolean) like when the whole job is finished doing its work, this will require you passing the JobParameter to some methods, but I don't think that would be hard in AS.

**Comment by thermatk** [comment]
@k0shk0sh done, testing here

**Comment by k0shk0sh** [comment]
great, here is a notification for you to test with :p

**Comment by thermatk** [comment]
thanks :+1: ticks as a clock, cleaning up and committing

**Comment by thermatk** [comment]
While you're testing, I think I can make the code more beautiful :rofl: 
duration pref could be long, no conversions needed in this case, etc

**Comment by k0shk0sh** [comment]
@thermatk you can do that for sure as long as we don't break existing behavior. I'm out, so I'll probably test it after I get back home. 
Thanks btw.
